src/charm.py

```diff
@@ -69,29 +71,65 @@ def _on_config_changed(self, event) -> None:
+        """Validate a string type: ipv4, ipv6 or invalid as IP address."""
+        try:
+            ip = ipaddress.ip_address(address)
+
+            if isinstance(ip, ipaddress.IPv4Address):
+                return "ipv4"
+            elif isinstance(ip, ipaddress.IPv6Address):
+                return "ipv6"
+        except ValueError:
+            return
+
+    def readlines_reverse(self, qfile):
+        with qfile:
+            qfile.seek(0, os.SEEK_END)
+            position = qfile.tell()
+            line = StringIO("")
+            while position >= 0:
+                qfile.seek(position)
+                next_char = qfile.read(1)
+                if next_char == "\n":
+                    yield line.getvalue()[::-1]
+                    line = StringIO("")
+                else:
+                    line.write(next_char)
+                position -= 1
+            yield line.getvalue()[::-1]
+
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Handle the report-visits-by-ip action."""
-        results = self._report_visits_by_ip(event)
+        """Handle the report-visits-by-ip action.
+
+        Args:
+            event: the Juju action event fired when the action executes.
+        """
+        results = self._report_visits_by_ip()
         event.set_results({"ips": str(results)})
 
-    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+    def filter_lines(self, file) -> list[str]:
+        for line in self.readlines_reverse(file):
+            line = line.split()
+            if len(line) > 1 and datetime.strptime(
+                line[3].lstrip("[").rstrip("]"), "%d/%b/%Y:%H:%M:%S"
+            ) > (datetime.now() - timedelta(minutes=20)):
+                yield line
+            else:
+                break
+
+    def get_ip(self, log_list) -> list[str]:
+        line_list = list(self.filter_lines(log_list))
+        for line in line_list:
+            yield str(line[0])
+
+    def _report_visits_by_ip(self) -> list[(int, str)]:
         """Report requests to nginx grouped and ordered by IP and report action result."""
         container = self.unit.get_container(CONTAINER_NAME)
         log_path = "/var/log/nginx/access.log"
-        with container.pull(log_path) as log_file:
-            list_to_search = log_file.readlines()
-        ip_regex = r"[0-9]+(?:\.[0-9]+){3}"
-        ip_list = []
-        current_datetime = datetime.now()
-        start_datetime = current_datetime - timedelta(minutes=20)
-        for line in list_to_search:
-            line = line.split()
-            log_datetime = datetime.strptime(line[3].lstrip("[").rstrip("]"), "%d/%b/%Y:%H:%M:%S")
-            if log_datetime >= start_datetime and re.search(ip_regex, line[0]):
-                ip_list.append(str(line[0]))
-        return sorted(
-            [(key, len(list(group))) for key, group in groupby(ip_list)], key=lambda t: t[1]
-        )
+        ip_list = list(self.get_ip(container.pull(log_path)))
```

Comment:
It would be better if this was a few statements using the functions sequentially to limit the coupling between the functions, e.g.:

* Get an iterator that yields lines from the log file in reverse order
* Possibly filter that for lines we don't want to process (e.g., empty lines, lines without IP addresses, lines without dates)
* Apply a takewhile to stop taking lines once a date is exceeded
* Apply a transformation from line to IP address or even just a reduction to a `Counter` using: https://docs.python.org/3/library/functools.html#functools.reduce, the latter would make this code quite efficient

This simplifies testing of the individual functions